modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentManagerImpl.java

```
@@ -0,0 +1,410 @@
+        clusterService.messagingService().addMessageHandler(DeployUnitMessageTypes.class,
+                (message, senderConsistentId, correlationId) -> {
+
+                    if (message instanceof DeployUnitRequest) {
+                        processDeployRequest((DeployUnitRequest) message, senderConsistentId, correlationId);
+                    } else if (message instanceof UndeployUnitRequest) {
+                        processUndeployRequest((UndeployUnitRequest) message, senderConsistentId, correlationId);
+                    }
+                });
+    }
+
+    private void processDeployRequest(DeployUnitRequest executeRequest, String senderConsistentId, long correlationId) {
+        doDeploy(executeRequest).whenComplete((success, throwable) -> {
+            DeployUnitResponseBuilder builder = DeployUnitResponseImpl.builder();
+            if (throwable != null) {
+                builder.error(throwable);
+            }
+            clusterService.messagingService().respond(senderConsistentId,
+                    builder.build(), correlationId);
+        });
+    }
+
+    private void processUndeployRequest(UndeployUnitRequest executeRequest, String senderConsistentId, long correlationId) {
+        try {
+            Path unitPath = unitsFolder
+                    .resolve(executeRequest.id())
+                    .resolve(executeRequest.version());
+
+            Files.walkFileTree(unitPath, new SimpleFileVisitor<>() {
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    Files.delete(file);
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                    Files.delete(dir);
+                    return FileVisitResult.CONTINUE;
+                }
+            });
```

```suggestion
            IgniteUtils.deleteIfExists(unitPath);
```

---

`postVisitDirectory`'s second argument should be rethrown if it's not null or else it will be silently swallowed